packages/dataviews/src/view-list.tsx

```diff
@@ -142,23 +205,136 @@ function ListItem( {
 						</HStack>
 					</CompositeItem>
 				</div>
+				{ actions?.length > 0 && (
+					<HStack
+						spacing={ 1 }
+						justify="flex-end"
+						className="dataviews-view-list__item-actions"
+						style={ {
+							flexShrink: '0',
+							width: 'auto',
+						} }
+					>
+						{ primaryAction && !! primaryAction.RenderModal && (
+							<div role="gridcell">
+								<CompositeItem
+									store={ store }
+									render={
+										<Button
+											label={ primaryAction.label }
+											icon={ primaryAction.icon }
+											isDestructive={
+												primaryAction.isDestructive
+											}
+											size="compact"
+											onClick={ () =>
+												setIsModalOpen( true )
+											}
+										/>
+									}
+								>
+									{ isModalOpen && (
+										<ActionModal
+											action={ primaryAction }
+											items={ [ item ] }
+											closeModal={ () =>
+												setIsModalOpen( false )
+											}
+											onActionStart={ () => {} }
+											onActionPerformed={ () => {} }
+										/>
+									) }
+								</CompositeItem>
+							</div>
+						) }
+						{ primaryAction && ! primaryAction.RenderModal && (
+							<div role="gridcell" key={ primaryAction.id }>
+								<CompositeItem
+									store={ store }
+									render={
+										<Button
+											label={ primaryAction.label }
+											icon={ primaryAction.icon }
+											isDestructive={
+												primaryAction.isDestructive
+											}
+											size="compact"
+											onClick={ () =>
+												primaryAction.callback( [
+													item,
+												] )
+											}
+										/>
+									}
+								/>
+							</div>
+						) }
+						<div role="gridcell">
+							<DropdownMenu
+								trigger={
+									<CompositeItem
+										store={ store }
+										render={
+											<Button
+												size="compact"
+												icon={ moreVertical }
+												label={ __( 'Actions' ) }
+												disabled={ ! actions.length }
+												onKeyDown={ ( event: {
```

Comment:
cc @WordPress/gutenberg-components for awareness: I had to implement the `onKeyDown` event in the trigger `Button` to prevent the arrow down from opening the menu. That's not necessarily when using Ariakit directly, see example at https://github.com/ariakit/ariakit/issues/3768